sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+# Behavior for optional kernel features
+
+This design document describes the changes that are needed in DPC++ in order to
+conform to the SYCL 2020 specification regarding the behavior of applications
+that use optional kernel features.  An optional kernel feature is any feature
+that is supported by some devices and not by others.  For example, not all
+devices support 16-bit floating point operations, so the `sycl::half` data type
+is an optional kernel feature.  Some DPC++ extensions like AMX are also
+optional kernel features.
+
+The requirements for this design come mostly from the SYCL 2020 specification
+[section 5.7 "Optional kernel features"][1] but they also encompass the C++
+attribute `[[sycl::requires()]]` that is described in [section 5.8.1 "Kernel
+attributes"][2] and [section 5.8.2 "Device function attributes"][3].
+
+[1]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features>
+[2]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes>
+[3]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_function_attributes>
+
+
+## Requirements
+
+There are several categories of requirements covered by this design:
+
+* The front-end compiler must issue a diagnostic in some cases when a kernel or
+  device function uses an optional feature.  However, the front-end compiler
+  must **not** generate a diagnostic in other cases.
+
+* The runtime must raise an exception when a kernel using optional features
+  is submitted to a device that does not support those features.  This
+  exception must be raised synchronously from the kernel invocation command
+  (e.g. `parallel_for()`).
+
+* The runtime must not raise an exception (or otherwise fail) merely due to
+  speculative compilation of a kernel for a device, when the application does
+  not specifically submit the kernel to that device.
+
+### Diagnostics from the front-end compiler
+
+By "front-end compiler", we mean the DPC++ compiler which parses DPC++ source
+code, not the JIT compiler that translates SPIR-V into native code.
+
+In general, the front-end compiler does not know which kernels the application
+will submit to which devices.  Therefore, the front-end compiler does not
+generally know which optional features a kernel can legally use.  Thus, in
+general, the front-end compiler must not issue any diagnostic simply because a
+kernel uses an optional feature.
+
+The only exception to this rule occurs when the application uses the C++
+attribute `[[sycl::requires()]]`.  When the application decorates a kernel or
+device function with this attribute, it is an assertion that the kernel or
+device function is allowed to use only those optional features which are listed
+by the attribute.  Therefore, the front-end compiler must issue a diagnostic if
+the kernel or device function uses any other optional kernel features.
```

Comment:
What does "kernel or device function uses any other optional kernel features" mean? Does it mean that diagnostic should be emitted if functions called from this kernel/device function are decorated with aspects not mentioned in parent list?